File: frontend/src/lib/matching.ts
import { civicBodies } from "@/data/civic-bodies";
import { getCityMeta } from "@/data/locations";
import type {
  Address,
  CivicBody,
  Department,
  MatchedCivicBody,
} from "@/types";

function haversineKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number
) {
  const toRad = (value: number) => (value * Math.PI) / 180;
  const earthRadius = 6371;
  const dLat = toRad(lat2 - lat1);
  const dLon = toRad(lon2 - lon1);
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(toRad(lat1)) * Math.cos(toRad(lat2)) * Math.sin(dLon / 2) ** 2;

  return 2 * earthRadius * Math.asin(Math.sqrt(a));
}

function scoreBody(
  body: CivicBody,
  address: Address,
  department?: Department
): MatchedCivicBody {
  let score = 0;
  const reasons: string[] = [];
  const cityMeta = getCityMeta(address.city, address.state);
  const prefix = address.pincode.slice(0, 3);

  if (department && body.departments.includes(department)) {
    score += 30;
    reasons.push("Handles this complaint type");
  }

  if (body.pincode === address.pincode) {
    score += 100;
    reasons.push("Exact PIN code match");
  } else if (body.pincodePrefixes.includes(prefix)) {
    score += 55;
    reasons.push("Same postal circle");
  }

  if (body.city.toLowerCase() === address.city.trim().toLowerCase()) {
    score += 40;
    reasons.push("Same city");
  } else if (body.state.toLowerCase() === address.state.trim().toLowerCase()) {
    score += 12;
    reasons.push("Same state");
  }

  let distanceKm: number | null = null;
  if (cityMeta) {
    distanceKm = Number(
      haversineKm(cityMeta.lat, cityMeta.lng, body.lat, body.lng).toFixed(1)
    );
    score += Math.max(0, 25 - Math.min(25, distanceKm));
    if (distanceKm <= 8) {
      reasons.push(`About ${distanceKm} km from your locality`);
    }
  }

  return { body, score, distanceKm, matchReasons: reasons };
}

export function findNearestBodies(
  address: Address,
  department?: Department,
  limit = 4
): MatchedCivicBody[] {
  const ranked = civicBodies
    .filter((body) => (department ? body.departments.includes(department) : true))
    .map((body) => scoreBody(body, address, department))
    .sort((a, b) => {
      if (b.score !== a.score) return b.score - a.score;
      const distanceA = a.distanceKm ?? 9999;
      const distanceB = b.distanceKm ?? 9999;
      return distanceA - distanceB;
    });

  const withLocalSignal = ranked.filter((item) => item.score >= 40);
  const pool = withLocalSignal.length > 0 ? withLocalSignal : ranked;

  const uniqueByType = new Map<string, MatchedCivicBody>();
  for (const match of pool) {
    if (!uniqueByType.has(match.body.type)) {
      uniqueByType.set(match.body.type, match);
    }
  }

  return [...uniqueByType.values()].slice(0, limit);
}

export function findBestBodyForDepartment(
  address: Address,
  department: Department
) {
  return findNearestBodies(address, department, 1)[0] ?? null;
}

export function assignHomeCivicBodies(address: Address) {
  const municipal = findBestBodyForDepartment(address, "roads");
  const electricity = findBestBodyForDepartment(address, "electricity");
  const water = findBestBodyForDepartment(address, "water");
  const traffic = findBestBodyForDepartment(address, "traffic");

  const matches = [municipal, electricity, water, traffic].filter(
    (item): item is MatchedCivicBody => Boolean(item)
  );

  const seen = new Set<string>();
  return matches.filter((item) => {
    if (seen.has(item.body.id)) return false;
    seen.add(item.body.id);
    return true;
  });
}

export function formatAddress(address: Address) {
  return [address.line1, address.area, address.city, address.state, address.pincode]
    .filter(Boolean)
    .join(", ");
}

File: android/app/src/main/java/in/civicconnect/app/data/Matching.kt
package `in`.civicconnect.app.data

import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sin
import kotlin.math.sqrt

private fun haversineKm(lat1: Double, lon1: Double, lat2: Double, lon2: Double): Double {
    val earth = 6371.0
    val dLat = Math.toRadians(lat2 - lat1)
    val dLon = Math.toRadians(lon2 - lon1)
    val a = sin(dLat / 2).pow(2) +
        cos(Math.toRadians(lat1)) * cos(Math.toRadians(lat2)) * sin(dLon / 2).pow(2)
    return 2 * earth * asin(sqrt(a))
}

private fun scoreBody(body: CivicBody, address: Address, department: String?): MatchedCivicBody {
    var score = 0
    val reasons = mutableListOf<String>()
    val coords = cityCoords[address.city] ?: cityCoords.entries.find {
        it.key.equals(address.city, true)
    }?.value
    val prefix = address.pincode.take(3)

    if (department != null && body.departments.contains(department)) {
        score += 30
        reasons += "Handles this complaint type"
    }
    if (body.pincode == address.pincode) {
        score += 100
        reasons += "Exact PIN code match"
    } else if (body.pincodePrefixes.contains(prefix)) {
        score += 55
        reasons += "Same postal circle"
    }
    if (body.city.equals(address.city.trim(), true)) {
        score += 40
        reasons += "Same city"
    } else if (body.state.equals(address.state.trim(), true)) {
        score += 12
        reasons += "Same state"
    }

    var distance: Double? = null
    if (coords != null) {
        distance = String.format("%.1f", haversineKm(coords.first, coords.second, body.lat, body.lng)).toDouble()
        score += maxOf(0, 25 - min(25.0, distance).toInt())
        if (distance <= 8) reasons += "About $distance km from your locality"
    }
    return MatchedCivicBody(body, score, distance, reasons)
}

fun findNearestBodies(address: Address, department: String? = null, limit: Int = 4): List<MatchedCivicBody> {
    val ranked = civicBodies
        .filter { department == null || it.departments.contains(department) }
        .map { scoreBody(it, address, department) }
        .sortedWith(compareByDescending<MatchedCivicBody> { it.score }.thenBy { it.distanceKm ?: 9999.0 })
    val pool = ranked.filter { it.score >= 40 }.ifEmpty { ranked }
    val unique = linkedMapOf<String, MatchedCivicBody>()
    pool.forEach { if (!unique.containsKey(it.body.type)) unique[it.body.type] = it }
    return unique.values.take(limit)
}

fun findBestBodyForDepartment(address: Address, department: String) =
    findNearestBodies(address, department, 1).firstOrNull()

fun assignHomeCivicBodies(address: Address): List<MatchedCivicBody> {
    val matches = listOfNotNull(
        findBestBodyForDepartment(address, "roads"),
        findBestBodyForDepartment(address, "electricity"),
        findBestBodyForDepartment(address, "water"),
        findBestBodyForDepartment(address, "traffic")
    )
    return matches.distinctBy { it.body.id }
}
